web/src/lib/stores/assets.store.ts

```diff
@@ -1,38 +1,247 @@
-  updateBucket: (bucket: string, actualBucketHeight: number) => number;
-  cancelBucket: (bucket: AssetBucket) => void;
+export type AssetStoreOptions = GetAssetCountByTimeBucketDto;
 
-  // asset
-  removeAsset: (assetId: string) => void;
-  updateAsset: (assetId: string, isFavorite: boolean) => void;
+export interface Viewport {
+  width: number;
+  height: number;
+}
 
-  // asset navigation
-  getNextAssetId: (assetId: string) => Promise<string | null>;
-  getPreviousAssetId: (assetId: string) => Promise<string | null>;
+interface AssetLookup {
+  bucket: AssetBucket;
+  bucketIndex: number;
+  assetIndex: number;
+}
 
-  // store
-  subscribe: (run: (value: AssetGridState) => void, invalidate?: (value?: AssetGridState) => void) => () => void;
+export class AssetBucket {
+  /**
+   * The DOM height of the bucket in pixel
+   * This value is first estimated by the number of asset and later is corrected as the user scroll
+   */
+  bucketHeight!: number;
+  bucketDate!: string;
+  assets!: AssetResponseDto[];
+  cancelToken!: AbortController | null;
+  position!: BucketPosition;
 }
 
-export function createAssetStore(): AssetStore {
-  const store = new AssetGridState();
-
-  return {
-    init: store.init.bind(store),
-    loadBucket: store.loadBucket.bind(store),
-    updateBucket: store.updateBucket.bind(store),
-    cancelBucket: store.cancelBucket.bind(store),
-    removeAsset: store.removeAsset.bind(store),
-    updateAsset: store.updateAsset.bind(store),
-    getNextAssetId: store.getNextAssetId.bind(store),
-    getPreviousAssetId: store.getPreviousAssetId.bind(store),
-    subscribe: store.subscribe,
-  };
+const THUMBNAIL_HEIGHT = 235;
+
+export class AssetStore {
+  private store$ = writable(this);
+  private assetToBucket: Record<string, AssetLookup> = {};
+
+  timelineHeight = 0;
+  buckets: AssetBucket[] = [];
+  assets: AssetResponseDto[] = [];
+
+  constructor(private options: AssetStoreOptions) {
+    this.store$.set(this);
+  }
+
+  subscribe = this.store$.subscribe;
+
+  async init(viewport: Viewport) {
+    const { data } = await api.assetApi.getAssetCountByTimeBucket({
+      getAssetCountByTimeBucketDto: { ...this.options, withoutThumbs: true },
+    });
+
+    this.buckets = data.buckets.map((bucket) => {
+      const unwrappedWidth = (3 / 2) * bucket.count * THUMBNAIL_HEIGHT * (7 / 10);
+      const rows = Math.ceil(unwrappedWidth / viewport.width);
+      const height = rows * THUMBNAIL_HEIGHT;
+
+      return {
+        bucketDate: bucket.timeBucket,
+        bucketHeight: height,
+        assets: [],
+        cancelToken: null,
+        position: BucketPosition.Unknown,
+      };
+    });
+
+    this.timelineHeight = this.buckets.reduce((acc, b) => acc + b.bucketHeight, 0);
+
+    this.emit(false);
+
+    let height = 0;
+    for (const bucket of this.buckets) {
+      if (height < viewport.height) {
+        // TODO use viewport.width + wrapping in estimate
```

Comment:
> Not sure we need this TODO. Google [says](https://medium.com/google-design/google-photos-45b714dfbed1) that height estimation above (`const unwrappedWidth = (3 / 2) * bucket.count * THUMBNAIL_HEIGHT * (7 / 10); ...`) doesn't need to be close to accurate:
> 
> > You may be asking yourself how could this possibly be accurate? The truth is it isn’t, not even close.
> > It’s fortunate that I initially overthought this part of the problem (for reasons I’ll explain in the Layout section) but it turns out that you don’t need the estimate to be very good (and for large numbers of photos it is often wrong by tens of thousands of pixels). The only important thing is that the estimate is vaguely representative so that the scrollbar keeps the appearance of accuracy.

After additional thinking my guess is that you may be reffering to a "one-photo-a-day" case. I think that's a quite rare situation. We can just delegate this task to the `IntersectionObserver`

Reply:
Ahh, makes sense. Maybe it is fine as is then.

Reply:
I removed it. I think the issue is much more common during development than it would be in production. It is more jarringly wrong when you only have a handful of photos are they are all on different days. Sounds like I just need better test data :smile: